src/container/include/container/container_parsing.hpp

```diff
@@ -0,0 +1,107 @@
+#ifndef CLIPPER_CONTAINER_PARSING_HPP
+#define CLIPPER_CONTAINER_PARSING_HPP
+
+#include <cstring>
+#include <vector>
+
+#include <container/datatypes.hpp>
+
+namespace clipper {
+
+namespace container {
+
+template <typename T>
+struct input_parser_type {
+  static const bool is_string_parser = false;
+};
+
+template <>
+struct input_parser_type<char> {
+  static const bool is_string_parser = true;
+};
+
+template <typename D>
+class InputParser {
+ public:
+  InputParser(std::vector<D>& buffer) : buffer_(buffer) {}
+  /**
+   * Exposes a buffer into which request data should be
+   * read directly from a socket. The buffer contents
+   * can then be parsed via `get_inputs`
+   */
+  std::vector<D>& get_data_buffer(long min_size_bytes) {
+    resize_if_necessary(buffer_, min_size_bytes);
+    return buffer_;
+  }
+
+  const std::vector<Input<D>> get_inputs(const std::vector<int>& input_header,
+                                         long input_content_size) {
+    if (input_parser_type<D>::is_string_parser) {
+      return get_string_inputs(input_header, input_content_size);
+    } else {
+      return get_primitive_inputs(input_header, input_content_size);
+    }
+  }
+
+ private:
+  std::vector<D>& buffer_;
+
+  void resize_if_necessary(std::vector<D>& buffer, long required_buffer_size) {
+    if (static_cast<long>((buffer.size() * sizeof(D))) < required_buffer_size) {
+      buffer.reserve((2 * required_buffer_size) / sizeof(D));
+    }
+  }
+
+  const std::vector<Input<D>> get_primitive_inputs(
+      const std::vector<int>& input_header, long input_content_size) {
+    // For a prediction request containing `n` inputs, there are `n - 1` split
+    // indices
+    int num_splits = input_header[1] - 1;
+    std::vector<Input<D>> inputs;
+    int prev_split = 0;
+    // Iterate from the beginning of the input data content
+    // (the first two elements of the header are metadata)
+    for (int i = 2; i < num_splits + 2; i++) {
```

Comment:
Added the following comment and variable definitions to make things clearer:

```
    // The first two elements of the header indicate the type of input data
    // and the number of content split indices. Therefore, actual splits
    // begin at header index 2 and continue through header index `num_splits` +
    // 2
    int splits_begin_index = 2;
    int splits_end_index = num_splits + 2;
```
The loop is now
```
for (int i = splits_begin_index; i < splits_end_index; ++i) {
    ....
}
```